Added Q&A and some minor fixes (#207)
pnp/blog: content/office-add-ins-community-call/2022-05-11/index.md

## Actions

* Join the Teams call and give us feedback on what you need from samples – [Office Add-ins samples discussion](https://aka.ms/officeaddinsamples-call) – June 15, 8:00 am PT \| aka.ms/officeaddinsamples-call
* Earn contributor badges - sign up for the PnP Recognition program. - <https://pnp.github.io/recognitionprogram/>
* Follow channels in twitter – for call agendas and summaries, release announcements,…
    * <https://twitter.com/microsoft365dev>
    * <https://twitter.com/m365pnp>
* Register for the [Microsoft 365 Developer Program](https://aka.ms/m365/devprogram) and get a free E5 developer tenant with instant availability and other assets.
* Next community call – June 8th at 08:00am PT - <https://aka.ms/officeaddinscommunitycall>

## Q&A (Question & Answers)

> **I'd love to follow-up on this feature request: [Office JS Save and Close methods for Word, Excel, and PowerPoint](https://techcommunity.microsoft.com/t5/microsoft-365-developer-platform/office-js-save-and-close-methods-for-word-excel-and-powerpoint/idi-p/2800214). Apologies for my persistence, but this is literarily the only blocking issue for us to move from the old add-ins to the Office web Add-ins platform. We would really really appreciate if you could push this into the roadmap (and we represent many clients with many employees, so do not take the 20 upvotes at face value.) We represent hundreds of users and the deployments of the old add-ins and updates are not ideal. Thank you so much.**

Reviewer: ```suggestion
**I'd love to follow-up on this feature request: [Office JS Save and Close methods for Word, Excel, and PowerPoint](https://techcommunity.microsoft.com/t5/microsoft-365-developer-platform/office-js-save-and-close-methods-for-word-excel-and-powerpoint/idi-p/2800214). Apologies for my persistence, but this is literally the only blocking issue for us to move from the old add-ins to the Office web Add-ins platform. We would really appreciate if you could push this into the roadmap. We represent hundreds of users and the deployments of the old add-ins and updates are not ideal. Thank you so much.**
```

> Thanks for your persistence! We're sorry you are being blocked by this. Unfortunately we can't add this to the roadmap at this time, but we will continue to keep an eye on the request. 

Reviewer: ```suggestion
Thanks for your persistence! We're sorry you're blocked by this. Unfortunately, we can't add this to the roadmap at this time, but we'll continue to keep an eye on the request. Consider asking the [Office JS developer community](https://stackoverflow.com/questions/tagged/office-js) on Stack Overflow how others have addressed this issue.
```
Consider updating the question on TechCommunity, too!

> As a workaround you may want to explore continuing to use your COM add-in for full functionality while adding partial functionality in the web add-in for cross-platform support. We recommend exploring this article for more information. [Make your Office Add-in compatible with an existing COM add-in](https://docs.microsoft.com/office/dev/add-ins/develop/make-office-add-in-compatible-with-existing-com-add-in)

Reviewer: ```suggestion
As a workaround you may want to explore continuing to use your COM add-in for full functionality while adding partial functionality in the web add-in for cross-platform support. For more information, we recommend exploring [Make your Office Add-in compatible with an existing COM add-in](https://docs.microsoft.com/office/dev/add-ins/develop/make-office-add-in-compatible-with-existing-com-add-in).
```

**Open the Outlook mobile API to allow compose scenarios to insert attachments in emails. Currently, third-party cloud storage companies cannot integrate directly into Outlook mobile. With this API, we would be able to click a button and insert files from private cloud storage providers into emails.**

> We'd like to know more about this scenario. Can you please create a feature request with more information at https://aka.ms/m365dev-suggestions? Thanks!

Reviewer: ```suggestion
We'd like to know more about this scenario. Can you please create a feature request with more information at [https://aka.ms/m365dev-suggestions](https://aka.ms/m365dev-suggestions)? Thanks!
```

**How can you communicate from a Word add-in to an Office JS task pane? For example, OnContentEnter, highlight one HTML control in the task pane.**

> We see this question was also [posted to the Q&A site](https://docs.microsoft.com/answers/questions/827167/how-to-call-office-js-function-hosted-in-taskpane.html). It seems like you should be able to use the [onSelectionChanged](https://docs.microsoft.com/javascript/api/word/word.contentcontrol?view=word-js-preview#word-word-contentcontrol-onselectionchanged-member) event. If not can you please follow up with more information on the Q&A site? Thanks!

Reviewer: ```suggestion
We see this question was also [posted to the Q&A site](https://docs.microsoft.com/answers/questions/827167/how-to-call-office-js-function-hosted-in-taskpane.html). You should be able to use the [onSelectionChanged](https://docs.microsoft.com/javascript/api/word/word.contentcontrol?view=word-js-preview#word-word-contentcontrol-onselectionchanged-member) event. If not, can you please follow up with more information on the Q&A site? Thanks!
```
Is this more like an issue? Should we point them to office-js repo instead?

> **During the last call, one of my questions was if it would be possible to programmatically let the compose window disappear after the user clicked on a custom "Encrypt and Send" button, as the mail already is encrypted. Then send the email as a MIME payload via the send message Graph API. You suggested to use smart alerts to handle the send event, however, this is not ideal for our use case. The on send event is triggered on each email, but we want the user to deliberately decide via the custom button if they want to encrypt the email or not. Also, encrypting "on the fly" via OnMessage events is not on option as not every mail should be encrypted. Also it does not work well with our custom encryption protocol. Hence, a feature request could be to programmatically discard the current composed mail, and subsequently close the compose form.**

Reviewer: ```suggestion
**During the last call, one of my questions was is possible to programmatically let the compose window disappear after the user clicks a custom "Encrypt and Send" button, as the mail already is encrypted. Then send the email as a MIME payload via the send message Graph API. You suggested to use smart alerts to handle the send event. However, this is not ideal for our use case. The on-send event is triggered on each email, but we want the user to deliberately decide via the custom button if they want to encrypt the email or not. Also, encrypting "on the fly" via OnMessage events is not on option as not every mail should be encrypted. Also, it does not work well with our custom encryption protocol. Hence, a feature request could be to programmatically discard the current composed mail, and subsequently close the compose form.**
```

> We are not planning to override or suppress the native send button, nor to expose a cancel composition API This can cause user confusion, and could be used maliciously in some conditions. The scenario can still be solved with smart alerts. The key here is to combine the following things.

Reviewer: ```suggestion
We are not planning to override or suppress the native send button, nor to expose a cancel composition API. This can cause user confusion and could be used maliciously in some conditions. The scenario can still be solved with smart alerts. The key here is to combine the following things.
```

* Custom properties
* LaunchEvents (attachmentsChanges particularly)
* Smart Alerts

> Pre-conditions: we cannot override the “send” button in Outlook. This is also not a recommended experience for Outlook users who expect to send an item using the native “send” option. So the ability to encrypt and send UEX-wise is not possible.

Reviewer: ```suggestion
Pre-conditions: we cannot override the “send” button in Outlook. This is also not a recommended experience for Outlook users who expect to send an item using the native “send” option. So, the ability to encrypt and send UEX-wise is not possible.
```

> 1. Indeed the user needs to take an explicit action to encrypt the content. So there needs to be an “Encrypt Message” button. The button needs to open a task pane and ask the user what they want to encrypt (should the add-in encrypt the message? the attachments? all?) Also the task pane can configure other settings or rules. For example a rule like “always encrypt when sending messages to this particular customer”. Those can be stored as a mailbox setting, so that afterwards the setting can be honored when running the onSend event.

Reviewer: ```suggestion
1. Indeed the user needs to take an explicit action to encrypt the content. So there needs to be an “Encrypt Message” button. The button needs to open a task pane and ask the user what they want to encrypt (should the add-in encrypt the message? the attachments? all?) Also, the task pane can configure other settings or rules. For example, a rule like “always encrypt when sending messages to this particular customer”. Those can be stored as a mailbox setting, so that afterwards the setting can be honored when running the on-send event.
```
This is good info. Is this a scenario you can add to the [on-send documentation](https://docs.microsoft.com/en-us/office/dev/add-ins/outlook/outlook-on-send-addins?tabs=windows)? Also, we apparently refer to these events collectively as "on-send" and only .OnSend when part of the actual event/method/property. Suggest changing references to on-send for consistency! (Or choose one or the other.)

2. When the user chooses the “Encrypt Message” button and configures what to encrypt, they can also create a custom property that serves as a flag for the onSend event. (i.e. pleaseEncrypt=true)
3. Then in the onSend event, the first thing to do is to check the pleaseEncrypt custom property. If it does not exist it means the event needs to do nothing and just send the email.

> 4. You can also avoid a heavy onSend management logic if you encrypt attachments as they are added using the attachment added event.

Reviewer: ```suggestion
4. To avoid a heavy onSend management logic, consider encrypting attachments as they are added using the AttachmentAdd event.
```
Consider including the link to the documentation.

> **Are you planning on upgrading the Visual Studio Web Add-in Templates to Asp.Net Core?**

Reviewer: ```suggestion
**Do you plant to upgrade the Visual Studio Web Add-in templates to ASP.NET Core?**
```

> We are looking into some updates to the Visual Studio Office Add-in templates, but don't have anything on the roadmap at this time. Can you let us know more about what you'd like to see in these templates by posting a request at https://aka.ms/m365dev-suggestions

Reviewer: ```suggestion
We are looking into some updates to the Visual Studio Office Add-in templates, but don't have anything on the roadmap at this time. Please let us know more about what you'd like to see in these templates by posting a request at [https://aka.ms/m365dev-suggestions](https://aka.ms/m365dev-suggestions).
```

**Has the PowerPoint API 1.3 been officially released to all platforms?**

The PowerPoint JavaScript API 1.3 requirement set is supported on Windows, Mac, and the web (but not iOS or mobile at this time.) See [PowerPoint JavaScript API requirement sets](https://docs.microsoft.com/javascript/api/requirement-sets/powerpoint/powerpoint-api-requirement-sets?view=powerpoint-js-preview) for complete details.

> **If the task pane is specified in the add in manifest.xml that is installed by an admin, it will always show in Excel files. Is there a way for Excel add-ins that are installed by an admin, to appear for a user or group only when the Excel file is created by the SaaS application, and not for existing or new Excel files created by the user.**

Reviewer: ```suggestion
**If the task pane is specified in the add in manifest.xml that is installed by an admin, it will always show in Excel files. Is there a way for Excel add-ins that are installed by an admin to appear for a user or group only when the Excel file is created by the SaaS application, and not for existing or new Excel files created by the user?**
```

> Once an add-in is deployed and it has **Show Taskpane** commands, those commands will appear in all Excel workbooks. You might be able to hide a custom tab if certain properties on the workbook are present. See the [visible property](https://docs.microsoft.com/javascript/api/office/office.tab?view=common-js-preview#office-office-tab-visible-member) for more information.

Reviewer: ```suggestion
The **Show Taskpane** commands appear in all Excel workbooks once an add-in is deployed. To hide a custom tab, certain properties on the workbook must be present. See the [visible property](https://docs.microsoft.com/javascript/api/office/office.tab?view=common-js-preview#office-office-tab-visible-member) for more information.
```

**The manifest validation service continues to fail for the Outlook add-in ItemSend event (see https://github.com/OfficeDev/Office-Addin-Scripts/issues/613). Our organization wants to use our build pipeline as we did in the past for admin deployed add-ins.**

> Thanks for pinging us on this. It could be that the PowerShell command is actually using the store validation that ACCEPTS launch event annotations, so this is probably failing erroneously. We'll investigate more and follow up on the original GitHub issue.

Reviewer: ```suggestion
Thanks for pinging us on this. It's possible that the PowerShell command is using the store validation that ACCEPTS launch event annotations causing it to fail erroneously. We'll investigate more and follow up on the original GitHub issue.
```
Consider adding the link to the original GitHub.

**Could AppSource submission provide a testing notes field with the submission that gets submitted with the app submission? Current testing notes always get lost between submissions! This might help get rid of the number one validation failures!**

You can attach a PDF to your submission that will stay for the lifetime of the validation even on subsequent submissions.

**Comment on AppSource: After submitting an update to an add-in, if the add-in is distributed to employees through Microsoft Admin Center, or Integrated apps, the add-in is never updated for the end-users (employees). Please could you review this issue?**

> This should not be the case, this could happen within 72 hours in the worst case. There are a few conditions (depending on the type of add-in) that requires an admin to approve the update (i.e. if your add-in was updated to support SSO, or event-based in Outlook). If you can provide more details we can definitely help you (what type of add-in is this). Please let us know more at https://stackoverflow.com or the [Q&A forum](https://aka.ms/office-addins-dev-questions).

Reviewer: ```suggestion
This should not be the case; this could happen within 72 hours in the worst case. Depending on the type of add-in, there are a few conditions that require an admin to approve the update (i.e. if your add-in was updated to support SSO, or event-based in Outlook). If you can provide more details we can definitely help you (what type of add-in is this). Please let us know more at [office-js on Stack Overflow](https://stackoverflow.com) or the [Q&A forum](https://aka.ms/office-addins-dev-questions).
```

**Regarding the add-in commands, there is noticeable lag between clicking a button and seeing the result. That lag does not exist when performing the exact same action from a task pane. Is there a reason for this, and a way around it?**

> When these commands are invoked, we start the runtime that is required to execute the action. This may result in a bit of a delay that you are experiencing. One way to work around this issue is to use the shared runtime in your add-in. The shared runtime starts the same runtime as the task pane and it continues to run in the background even when the task pane is closed. This avoids having to start the runtime every time a ribbon button is invoked for an action. For more information, see [Configure your Office Add-in to use a shared JavaScript runtime](https://docs.microsoft.com/office/dev/add-ins/develop/configure-your-add-in-to-use-a-shared-runtime).

Reviewer: ```suggestion
When these commands are invoked, we start the runtime that is required to execute the action. This may result in the delay that you are experiencing. A work around is to use the shared runtime in your add-in. The shared runtime starts the same runtime as the task pane and it continues to run in the background even when the task pane is closed. This avoids having to start the runtime every time a ribbon button is invoked for an action. For more information, see [Configure your Office Add-in to use a shared JavaScript runtime](https://docs.microsoft.com/office/dev/add-ins/develop/configure-your-add-in-to-use-a-shared-runtime).
```

**For Ribbon commands, is it ever going to be possible to have multiple add-ins appear in the same group?**

This is not something we are currently focused on. Please let us know more about this scenario at https://aka.ms/m365dev-suggestions.

**Will Office.actions.associate be recognized by the Intellisense in VS Code similar to CustomFunctions.associate?**

Yes, the associate API should be recognized by Intellisense. Please be sure to have the latest [office-js type library](https://www.npmjs.com/package/@types/office-js).

**Is the new Answers Q&A forum just for users to ask questions about existing functionality?**

> You can ask any technical question you need help with. We monitor both https://stackoverflow.com and the [Q&A forum](https://aka.ms/office-addins-dev-questions).

Reviewer: ```suggestion
You can ask any technical question you need help with. We monitor both [office-js questions on Stack Overflow](https://stackoverflow.com) and the [Q&A forum](https://aka.ms/office-addins-dev-questions).
```

**Is it recommended to post to this newer feedback portal, stick to the M365 developer tools, or post on GitHub?**

That is exactly what we're trying to determine. If we're going to ask customers to move yet again to a new forum, we want to be sure we're going to support it! Stay tuned, we'll get there. For the moment, we're using https://aka.ms/m365dev-sugggestions.

**Will Office.auth.getAccessToken ever return a Graph API compatible token?**

> We're investigating this. We ideally want to allow the same operations that we allow on EWS (like 14 services). But there's a lot of complexity to solve to allow this. We'll let you know down the road if we have updates on this. In the meantime, the recommended flow is that the server should get the Graph token using the [OAuth 2.0 On-Behalf-Of flow (OBO)](https://docs.microsoft.com/azure/active-directory/develop/v2-oauth2-on-behalf-of-flow).

Reviewer: ```suggestion
We're investigating this. Ideally, we want to allow the same operations that we allow on Exchange Web Services (EWS). But there's a lot of complexity to solve to allow this. We'll let you know if we have updates. In the meantime, the recommended flow is that the server should get the Graph token using the [OAuth 2.0 On-Behalf-Of flow (OBO)](https://docs.microsoft.com/azure/active-directory/develop/v2-oauth2-on-behalf-of-flow).
```
What are "14 services?"

**This might have come up before but since mobile add-ins only support the REST API, and REST is being decommissioned this year, any Outlook add-in on mobile that relies on REST API functionality with getAccessToken is going to break when REST API goes away right? There's no support for IdentityAPI 1.3 on mobile so no getAccessToken for Graph so no auth path it can take yeah?**

> In the case of mobile the alternative is to use EWS. You can also request an exception at https://aka.ms/RESTCheck that also applies to mobile. If your add-in is using REST before the November 2022 decommission date, it will be allowed to work until we end support for Office 2019.

Reviewer: ```suggestion
For mobile, the alternative is to use Exchange Web Services (EWS). You can also request an exception at https://aka.ms/RESTCheck that also applies to mobile. If your add-in is using REST before the November 2022 decommission date, it will be allowed to work until we end support for Office 2019.
```

**Does the add-in's getCallbackTokenAsync work on mobile to get a token that can be used against EWS on the server for authentication as the add-in user?**

Yes, this should work. We do have customers who have implemented a service middle tier to call EWS on environments not supporting REST (like Exchange 2013 on premise). You can also use that technique to use Microsoft Graph. This is all valid if your add-in is accessing Exchange 2013 on premise.

## References

* **AppSource Add-in Validation Improvements**

>     * Documentation - [Test your Office Add-in on Internet Explorer](https://docs.microsoft.com/office/dev/add-ins/testing/ie-11-testing)

Reviewer: ```suggestion
    * Documentation – [Test your Office Add-in on Internet Explorer](https://docs.microsoft.com/office/dev/add-ins/testing/ie-11-testing)
```

>     * Documentation - [Microsoft AppSource submission FAQ](https://docs.microsoft.com/office/dev/store/appsource-submission-faq)

Reviewer: ```suggestion
    * Documentation – [Microsoft AppSource submission FAQ](https://docs.microsoft.com/office/dev/store/appsource-submission-faq)
```

    * Issues – [Office Add-ins issues](https://github.com/OfficeDev/office-js/issues)

* **Community ideas**
    * Suggestions - [Microsoft 365 Developer Platform Ideas - Office Add-ins](https://aka.ms/m365dev-suggestions) \| [aka.ms/m365dev-suggestions](https://aka.ms/m365dev-suggestions)

>     * Issues - [OfficeDev/office-js/issues](https://github.com/OfficeDev/office-js/issues)

Reviewer: ```suggestion
    * Issues – [OfficeDev/office-js/issues](https://github.com/OfficeDev/office-js/issues)
```

>     * Forum - [Microsoft Ideas forum - Feedback Portal](https://feedbackportal.microsoft.com/feedback)

Reviewer: ```suggestion
    * Forum – [Microsoft 365 Developer Ideas forum](https://feedbackportal.microsoft.com/feedback)
```

Reviewer: For SEO, please consider adding the following at the bottom:
**Additional Resources**
[Microsoft 365 Developer Program](https://aka.ms/m365/devprogram)
[Microsoft 365 Dev Center](https://developer.microsoft.com/microsoft-365)
